usaspending_api/agency/v2/views/budget_function.py

```diff
@@ -0,0 +1,92 @@
+    """
+    Obtain the count of budget functions for a specific agency in a single
+    fiscal year based on whether or not that budget function has ever
+    been submitted in File B.
+    """
+
+    endpoint_doc = "usaspending_api/api_contracts/contracts/v2/agency/toptier_code/budget_function/budget_function.md"
+
+    def validate_request(self):
+        return None
+
+    def format_results(self, rows):
+        order = self.pagination.sort_order == "desc"
+        names = set([row["treasury_account__budget_function_title"] for row in rows])
+        budget_functions = [{"name": x} for x in names]
+        for item in budget_functions:
+            item["children"] = [
+                {
+                    "name": row["treasury_account__budget_subfunction_title"],
+                    "obligated_amount": row["obligated_amount"],
+                    "gross_outlay_amount": row["gross_outlay_amount"],
+                }
+                for row in rows
+                if item["name"] == row["treasury_account__budget_function_title"]
+            ]
+            item["obligated_amount"] = sum([x["obligated_amount"] for x in item["children"]])
+            item["gross_outlay_amount"] = sum([x["gross_outlay_amount"] for x in item["children"]])
+            item["children"] = sorted(item["children"], key=lambda x: x[self.pagination.sort_key], reverse=order)
+        budget_functions = sorted(budget_functions, key=lambda x: x[self.pagination.sort_key], reverse=order)
+        return budget_functions
+
+    @cache_response()
+    def get(self, request: Request, *args: Any, **kwargs: Any) -> Response:
+        results = self.format_results(list(self.get_budget_function_queryset()))[
+            self.pagination.lower_limit : self.pagination.upper_limit
+        ]
+        page_metadata = get_simple_pagination_metadata(len(results), self.pagination.limit, self.pagination.page)
+        return Response(
+            {
+                "toptier_code": self.toptier_code,
+                "fiscal_year": self.fiscal_year,
+                "limit": self.pagination.limit,
+                "results": results[: self.pagination.limit],
+                "messages": self.standard_response_messages,
+                "page_metadata": page_metadata,
+            }
+        )
+
+    def get_budget_function_queryset(self):
+        filters = [
+            Q(final_of_fy=True),
+            Q(treasury_account__funding_toptier_agency=self.toptier_agency),
+            Q(submission__reporting_fiscal_year=self.fiscal_year),
+        ]
+        if self.filter is not None:
+            filters.append(
+                Q(
+                    Q(treasury_account__budget_function_title__icontains=self.filter)
+                    | Q(treasury_account__budget_subfunction_title__icontains=self.filter)
+                )
+            )
+
+        results = (
+            (
+                FinancialAccountsByProgramActivityObjectClass.objects.filter(*filters).exclude(
```

Comment:
I think you are going to run into the same thing that was brought up on my PR. Using exclude here generated some funky SQL that @kbard noticed while reviewing my PR. For reference this is the change that Kirk suggested on my PR for program activity: https://github.com/fedspendingtransparency/usaspending-api/pull/2463/commits/97195a8585b2051e5b5079bcb0fd63b8fb91c0d3#diff-6dbc15851fa37562b438cd982343047cR36

Reply:
Also, here's a very detailed explanation of the problem if you're interested.  I think it's a Django bug.  They ask that you post in their user forum before actually posting a bug.

https://groups.google.com/forum/#!topic/django-users/jN93IsAGZAw